src/algebra/monoid_algebra.lean

```diff
@@ -373,6 +377,79 @@ end
+     smul_zero],
+end⟩
+
+/-- Note that if `k` is a `comm_semiring` then we have `smul_comm_class k k k` and so we can take
+`R = k` in the below. In other words, if the coefficients are commutative amongst themselves, they
+also commute with the algebra multiplication. -/
+instance smul_comm_class_self [smul_comm_class R k k] :
+  smul_comm_class R (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b,
+begin
+  ext m,
+  simp only [mul_apply, finsupp.sum, finset.smul_sum, smul_ite, mul_smul_comm, sum_smul_index',
+    implies_true_iff, eq_self_iff_true, coe_smul, ite_eq_right_iff, smul_eq_mul, pi.smul_apply,
+    mul_zero, smul_zero],
+end⟩
+
+instance smul_comm_class_symm_self [smul_comm_class k R k] :
+  smul_comm_class (monoid_algebra k G) R (monoid_algebra k G) :=
+⟨λ t a b, by { haveI := smul_comm_class.symm k R k, rw ← smul_comm, } ⟩
+
+/-- The functor `G ↦ monoid_algebra k G`, from the category of magmas to the category of non-unital,
+non-associative algebras over `k` is adjoint to the forgetful functor in the other direction. -/
+@[simps] def lift_magma {A : Type u₃}
+  [non_unital_non_assoc_semiring A] [module k A] [is_scalar_tower k A A] [smul_comm_class k A A] :
+  mul_hom G A ≃ non_unital_alg_hom k (monoid_algebra k G) A :=
+{ to_fun    := λ f,
+    { to_fun    := λ a, a.sum (λ m t, t • f m),
+      map_smul' :=  λ t' a,
+        begin
+          rw [finsupp.smul_sum, sum_smul_index'],
+          { simp_rw smul_assoc, },
+          { intros m, exact zero_smul k (f m), },
+        end,
+      map_mul'  := λ a₁ a₂,
+        begin
+          let g : G → k → A := λ m t, t • f m,
+          have h₁ : ∀ m, g m 0 = 0, { intros, exact zero_smul k (f m), },
+          have h₂ : ∀ m (t₁ t₂ : k), g m (t₁ + t₂) = g m t₁ + g m t₂, { intros, rw ← add_smul, },
+          simp_rw [finsupp.mul_sum, finsupp.sum_mul, smul_mul_smul, ← f.map_mul, mul_def,
+            sum_comm a₂ a₁, sum_sum_index h₁ h₂, sum_single_index (h₁ _)],
+        end,
+      .. lift_add_hom (λ x, (smul_add_hom k A).flip (f x)) },
+  inv_fun   := λ F, F.to_mul_hom.comp (of_magma k G),
+  left_inv  := λ f, by { ext m, simp only [non_unital_alg_hom.coe_mk, of_magma_apply,
+    non_unital_alg_hom.to_mul_hom_eq_coe, sum_single_index, function.comp_app, one_smul, zero_smul,
+    mul_hom.coe_comp, non_unital_alg_hom.coe_to_mul_hom], },
+  right_inv := λ F,
+    begin
+      ext a,
+      suffices : a.sum (λ m t, F (finsupp.single m t)) = F a,
+      { simp only [of_magma_apply, non_unital_alg_hom.coe_mk, mul_hom.coe_mk,
+          non_unital_alg_hom.to_mul_hom_eq_coe, linear_map.coe_mk, function.comp_app,
+          mul_hom.coe_comp, non_unital_alg_hom.coe_to_mul_hom, ← F.map_smul, smul_single,
+          smul_eq_mul, mul_one, this], },
+      let F' := (F : monoid_algebra k G →+[k] A).to_add_monoid_hom,
+      change a.sum (λ m t, F' (finsupp.single m t)) = F' a,
+      rw [← add_monoid_hom.map_finsupp_sum F', sum_single],
+    end, }
```

Comment:
I think this would become much shorter with a suitable `ext` lemma, which I think we should include as part of this PR:
```lean
variables {A : Type u₃}
  [non_unital_non_assoc_semiring A] [module k A] [is_scalar_tower k A A] [smul_comm_class k A A]

/-- A non_unital `k`-algebra homomorphism from `monoid_algebra k G` is uniquely defined by its
values on the functions `single a 1`. -/
lemma non_unital_alg_hom_ext ⦃φ₁ φ₂ : non_unital_alg_hom k (monoid_algebra k G) A⦄
  (h : ∀ x, φ₁ (single x 1) = φ₂ (single x 1)) : φ₁ = φ₂ :=
sorry
-- We should be able to adapt the proof of `alg_hom_ext` but many components of that are also missing


/-- See note [partially-applied ext lemmas]. -/
@[ext] lemma non_unital_alg_hom_ext' ⦃φ₁ φ₂ : non_unital_alg_hom k (monoid_algebra k G) A⦄
  (h : φ₁.to_mul_hom.comp (of_magma k G) = φ₂.to_mul_hom.comp (of_magma k G)) : φ₁ = φ₂ :=
non_unital_alg_hom_ext k $ mul_hom.congr_fun h
```

Reply:
Good point, I'll take this up later this morning.

Reply:
OK I've pushed https://github.com/leanprover-community/mathlib/pull/7932/commits/fbcff7e7de1db1521ad9ef0aef869ec63a740c36 to address this point.

I was a bit surprised that `distrib_mul_action_hom.to_linear_map` didn't already exist but it seems to be genuinely missing.